chainindexer.go

```diff
@@ -0,0 +1,497 @@
+}
+
+// addrIndexer provides a concurrent service for indexing the transactions of
+// target blocks based on the addresses involved in the transaction.
+type addrIndexer struct {
+	server          *server
+	started         int32
+	shutdown        int32
+	state           indexState
+	quit            chan struct{}
+	wg              sync.WaitGroup
+	addrIndexJobs   chan *indexBlockMsg
+	writeRequests   chan *writeIndexReq
+	progressLogger  *blockProgressLogger
+	currentIndexTip int64
+	chainTip        int64
+	sync.Mutex
+}
+
+// newAddrIndexer creates a new block address indexer.
+// Use Start to begin processing incoming index jobs.
+func newAddrIndexer(s *server) (*addrIndexer, error) {
+	_, chainHeight, err := s.db.NewestSha()
+	if err != nil {
+		return nil, err
+	}
+
+	_, lastIndexedHeight, err := s.db.FetchAddrIndexTip()
+	if err != nil && err != database.ErrAddrIndexDoesNotExist {
+		return nil, err
+	}
+
+	var state indexState
+	if chainHeight == lastIndexedHeight {
+		state = indexMaintain
+	} else {
+		state = indexCatchUp
+	}
+
+	ai := &addrIndexer{
+		server:          s,
+		quit:            make(chan struct{}),
+		state:           state,
+		addrIndexJobs:   make(chan *indexBlockMsg),
+		writeRequests:   make(chan *writeIndexReq, numCatchUpWorkers),
+		currentIndexTip: lastIndexedHeight,
+		chainTip:        chainHeight,
+		progressLogger: newBlockProgressLogger("Indexed addresses of",
+			adxrLog),
+	}
+	return ai, nil
+}
+
+// Start begins processing of incoming indexing jobs.
+func (a *addrIndexer) Start() {
+	// Already started?
+	if atomic.AddInt32(&a.started, 1) != 1 {
+		return
+	}
+	adxrLog.Trace("Starting address indexer")
+	a.wg.Add(2)
+	go a.indexManager()
+	go a.indexWriter()
+}
+
+// Stop gracefully shuts down the address indexer by stopping all ongoing
+// worker goroutines, waiting for them to finish their current task.
+func (a *addrIndexer) Stop() error {
+	if atomic.AddInt32(&a.shutdown, 1) != 1 {
+		adxrLog.Warnf("Address indexer is already in the process of " +
+			"shutting down")
+		return nil
+	}
+	adxrLog.Infof("Address indexer shutting down")
+	close(a.quit)
+	a.wg.Wait()
+	return nil
+}
+
+// IsCaughtUp returns a bool representing if the address indexer has
+// caught up with the best height on the main chain.
+func (a *addrIndexer) IsCaughtUp() bool {
+	a.Lock()
+	defer a.Unlock()
+	return a.state == indexMaintain
+}
+
+// indexManager creates, and oversees worker index goroutines.
+// indexManager is the main goroutine for the addresses indexer.
+// It creates, and oversees worker goroutines to index incoming blocks, with
+// the exact behavior depending on the current index state
+// (catch up, vs maintain). Completion of catch-up mode is always proceeded by
+// a gracefull transition into "maintain" mode.
+// NOTE: Must be run as a goroutine.
+func (a *addrIndexer) indexManager() {
+	if a.state == indexCatchUp {
+		adxrLog.Infof("Building up address index from height: %v to %v.",
```

Comment:
Heh, yeah you're right, was left over before I changed the logging message. Removed the colon.